src/sql/src/func.rs

```diff
@@ -2339,6 +2342,29 @@ pub static PG_CATALOG_BUILTINS: Lazy<BTreeMap<&'static str, Func>> = Lazy::new(|
         "pg_postmaster_start_time" => Scalar {
             params!() => UnmaterializableFunc::PgPostmasterStartTime => TimestampTz, 2560;
         },
+        "pg_relation_size" => Scalar {
+            params!(Any, String) => Operation::binary(|ecx, lhs, _| {
+                match ecx.scalar_type(&lhs) {
+                    ScalarType::Oid | ScalarType::RegClass => {}
+                    _ => sql_bail!("No function matches the given name and argument types. \
+                    You might need to add explicit type casts."),
+                }
+
+                Ok(HirScalarExpr::literal(Datum::Int64(0), ScalarType::Int64))
+            }) => Int64, 2332;
+            params!(Any) => Operation::unary(|ecx, e| {
+                match ecx.scalar_type(&e) {
+                    ScalarType::Oid | ScalarType::RegClass => {}
+                    _ => sql_bail!("No function matches the given name and argument types. \
+                    You might need to add explicit type casts."),
+                }
```

Comment:
Ditto.

Reply:
Thinking out loud. OID can’t be generically converted to RegClass. Even though regclass is what’s documented, I wonder if the param should be oid

Reply:
Sure it can!

```
benesch=# create table hiseth (a int);
CREATE TABLE
benesch=# select 'hiseth'::regclass::oid;
   oid   
---------
 1957795
(1 row)

benesch=# select 1957795::oid::regclass;
 regclass 
----------
 hiseth
(1 row)

benesch=# select 42::oid::regclass;
 regclass 
----------
 42
(1 row)

```

> Even though regclass is what’s documented, I wonder if the param should be oid

The great thing about pgcompat is we don't need to guess! The parameter should be exactly what `pg_proc` says it should be:

```
benesch=# select proargtypes::regtype[] from pg_proc where proname = 'pg_relation_size';
      proargtypes      
-----------------------
 [0:0]={regclass}
 [0:1]={regclass,text}
(2 rows)
```

So one overload should be (RegClass) and the other should be (RegClass, String). If that wasn't working for you, we must have the casting rules for RegClass wrong.

Reply:
```
benesch=# select * from pg_cast where castsource = 'oid'::regtype and casttarget = 'regclass'::regtype;
  oid  | castsource | casttarget | castfunc | castcontext | castmethod 
-------+------------+------------+----------+-------------+------------
 10074 |         26 |       2205 |        0 | i           | b
(1 row)
```

`pg_cast` says the cast from oid to regclass should be implicit.

Reply:
And we mistakenly have it as an "assignment" cast:

https://github.com/MaterializeInc/materialize/blob/f289ab89eb2b6f188f9193cb580fc1f8ab1200f8/src/sql/src/plan/typeconv.rs#L290-L292

If you apply this diff...

```diff
diff --git a/src/sql/src/plan/typeconv.rs b/src/sql/src/plan/typeconv.rs
index b24ad6db4c..b95df64254 100644
--- a/src/sql/src/plan/typeconv.rs
+++ b/src/sql/src/plan/typeconv.rs
@@ -287,9 +287,9 @@ static VALID_CASTS: Lazy<BTreeMap<(ScalarBaseType, ScalarBaseType), CastImpl>> =
         (Oid, Int32) => Assignment: CastOidToInt32(func::CastOidToInt32),
         (Oid, Int64) => Assignment: CastOidToInt32(func::CastOidToInt32),
         (Oid, String) => Explicit: CastOidToString(func::CastOidToString),
-        (Oid, RegClass) => Assignment: CastOidToRegClass(func::CastOidToRegClass),
-        (Oid, RegProc) => Assignment: CastOidToRegProc(func::CastOidToRegProc),
-        (Oid, RegType) => Assignment: CastOidToRegType(func::CastOidToRegType),
+        (Oid, RegClass) => Implicit: CastOidToRegClass(func::CastOidToRegClass),
+        (Oid, RegProc) => Implicit: CastOidToRegProc(func::CastOidToRegProc),
+        (Oid, RegType) => Implicit: CastOidToRegType(func::CastOidToRegType),
 
         // REGCLASS
         (RegClass, Oid) => Implicit: CastRegClassToOid(func::CastRegClassToOid),
```

then you should be able to define your functions in terms of only RegClass and the function selection logic will automatically cast OID inputs to RegClass.